dbms/src/Storages/Transaction/DecodingStorageSchemaSnapshot.h

```diff
@@ -100,10 +102,14 @@ struct DecodingStorageSchemaSnapshot
         // create pk related metadata if needed
         if (is_common_handle)
         {
-            const auto & primary_index_info = table_info_.getPrimaryIndexInfo();
-            for (size_t i = 0; i < primary_index_info.idx_cols.size(); i++)
+            /// we will not update the IndexInfo except Rename DDL.
+            /// When the add column / drop column action happenes, the offset of each column may change
+            /// Thus, we should not use offset to get the column we want,
+            /// but use to compare the column name to get the column id.
+            const auto & primary_index_cols = table_info_.getPrimaryIndexInfo().idx_cols;
+            for (const auto & col : primary_index_cols)
             {
-                auto pk_column_id = table_info_.columns[primary_index_info.idx_cols[i].offset].id;
+                auto pk_column_id = column_name_id_map[col.name];
```

Comment:
Maybe add a simple unittest for this test case:

* Parse tidb table info from a json string
* Get the `DecodingStorageSchemaSnapshot` by using `getDecodingStorageSchemaSnapshot` from tidb table info
* Check the member `pk_column_ids`, `pk_pos_map` etc